packages/lit-html/src/directives/repeat.ts

```diff
@@ -0,0 +1,445 @@
+ *
+ * Note that if a `keyFn` is provided, strict key-to-DOM mapping is maintained,
+ * meaning previous DOM for a given key is moved into the new position if
+ * needed, and DOM will never be reused with values for different keys (new DOM
+ * will always be created for new keys). This is generally the most efficient
+ * way to use `repeat` since it performs minimum unnecessary work for insertions
+ * and removals.
+ *
+ * IMPORTANT: If providing a `keyFn`, keys *must* be unique for all items in a
+ * given call to `repeat`. The behavior when two or more items have the same key
+ * is undefined.
+ *
+ * If no `keyFn` is provided, this directive will perform similar to mapping
+ * items to values, and DOM will be reused against potentially different items.
+ */
+
+class RepeatDirective extends Directive {
+  itemKeys?: unknown[];
+
+  constructor(part: Part) {
+    super();
+    if (!(part instanceof NodePart)) {
+      throw new Error('repeat can only be used in text bindings');
+    }
+  }
+
+  render<T>(
+    items: Iterable<T>,
+    keyFnOrTemplate: KeyFn<T> | ItemTemplate<T>,
+    template?: ItemTemplate<T>
+  ) {
+    let keyFn: KeyFn<T> | undefined;
+    if (template === undefined) {
+      template = keyFnOrTemplate;
+    } else if (keyFnOrTemplate !== undefined) {
+      keyFn = keyFnOrTemplate as KeyFn<T>;
+    }
+    this.itemKeys = [];
+    const values = [];
+    let index = 0;
+    for (const item of items) {
+      this.itemKeys[index] = keyFn ? keyFn(item, index) : index;
+      values[index] = template!(item, index);
+      index++;
+    }
+    return values;
+  }
+
+  update<T>(
+    containerPart: NodePart,
+    [items, keyFnOrTemplate, template]: [
+      Iterable<T>,
+      KeyFn<T> | ItemTemplate<T>,
+      ItemTemplate<T>
+    ]
+  ) {
+    // Old part & key lists are retrieved from the last update
+    // TODO: deal with directive being swapped out?
+    let oldParts = containerPart.__value as Array<NodePart | null>;
+    if (!oldParts) {
+      return this.render(items, keyFnOrTemplate, template);
+    }
+
+    let keyFn: KeyFn<T> | undefined;
+    if (template === undefined) {
+      template = keyFnOrTemplate;
+    } else if (keyFnOrTemplate !== undefined) {
+      keyFn = keyFnOrTemplate as KeyFn<T>;
+    }
+
+    const oldKeys = (this.itemKeys ??= []);
+
+    // New part list will be built up as we go (either reused from
+    // old parts or created for new keys in this update). This is
+    // saved in the above cache at the end of the update.
+    const newParts: NodePart[] = [];
+
+    // New value list is eagerly generated from items along with a
+    // parallel array indicating its key.
+    const newValues: unknown[] = [];
+    const newKeys: unknown[] = [];
+    let index = 0;
+    for (const item of items) {
+      newKeys[index] = keyFn ? keyFn(item, index) : index;
+      newValues[index] = template!(item, index);
+      index++;
+    }
```

Comment:
Done. I made a side-effect-free helper method though.

Reply:
Ugh, ok, but now `render` is not used on the client at all.  I think that's also a pretty weird pattern to implement a `render` that's _never_ used. My mental model is that `render` should be used for at least first render on both client & server.

Reply:
Because of the key array I think there's a choice here between it being side-effectful and always called, and being side-effect-free and not called on the client. I was thinking side-effect free would be better for cross-compilation. It'd be nice to be able to just drop `@soyCompatible` on here and have it work, though admittedly this isn't soy compatible because of the local var mutation, and we need to special-case repeat() anyway to be idom compatible...

Reply:
Agree, although there's a 3rd option which neither of us probably like, which is to do the loop twice, once in update to get/store the keys, and another in render to get the values. That would be the pure/semantic way to do it, at the cost of two loops.

One other downside to having directives where update never calls render is that we're not actually going to test the render implementation, at least in the client-side tests (note my related [comment on styleMap](https://github.com/Polymer/lit-html/pull/1218#discussion_r488317095) where this led to a bug). Yes, once we get the SSR tests in place, _we'll_ have good test coverage, but again this is evidence to me that it's pretty easy to accidentally write non-SSR compatible/correct code, where I was hoping we'd get a much stronger guarantee.

Reply:
I don't disagree, I just don't have a better option in my head yet. I don't want to do require a side-effectful `render()` or `super.update()` (that calls `render()`) to require the call to `render()`.

In LitElement if you don't call `super.update()` your element basically won't work, which is what effectively forces the call to `render()`. Of course an element could return nothing from `render()` and do all its rendering in `update()` or `updated()` and have a bad render in SSR. So there's no guarantee, even if LitElement pushes harder to the right thing. The equivalent would be something like `Directive.update()` setting some flag that it was called. I don't really want to require that directives extend from Directive though, even if it's what we usually document.